### PR TITLE
mrc-6062 Skeleton app

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch implements a basic express server that doesn't do anything except return package version. It includes:
- build and run scripts
- basic code structure: routes, controller, logging
- unit and integration tests (currently run from the same script, but these could be separated later).
- lint
- docker build
- basic github workflows

The next branch will include some example datasets and returning tile data. 

To do, in subsequent tickets: coverage and error handling

### CODE REVIEW

File: src/logging.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Application, Request, Response } from "express";
+import morgan from "morgan";
+import { Dict } from "./types";
+
+export const initialiseLogging = (app: Application) => {
```
Comment (EmmaLRussell) [inline src/logging.ts:5]: This format largely stolen from WODIN.

File: .github/workflows/test-and-push.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+          node-version: 22.x
+      - name: Install dependencies
+        run: npm ci
+      - name: Login to GHCR (GitHub Packages)
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build docker image
```
Comment (absternator) [inline .github/workflows/test-and-push.yml:27]: we should probably be testing code first before pushing? because then we create bad images

**Review by absternator — COMMENTED** [review]
Looks great!!! config is great.. just left  a few comments

File: .github/workflows/test-and-push.yml (hunk 2 above)
Comment (EmmaLRussell) [reply .github/workflows/test-and-push.yml:27]: This just builds and pushes the image for the SHA tag though, and then we run that for the integration tests. The branch tag only gets pushed once the tests pass, so branch tag should always be good. We've used this pattern elsewhere.

File: src/configReader.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import * as path from "path";
+
+export class ConfigReader {
+    rootDir: string;
+
+    constructor(rootDir: string) {
+        this.rootDir = rootDir;
+    }
+
+    readConfigFile(...filePath: string[]): object | null {
```
Comment (david-mears-2) [inline src/configReader.ts:11]: if the function / class is for reading config files, should it be allowed to have parameters? I think the filePath could be something known by ConfigReader. are we expecting there to be other config files to read?

File: src/controllers/indexController.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { Request, Response } from "express";
+
+export class IndexController {
+    static getIndex = (_req: Request, res: Response) => {
+        const version = process.env.npm_package_version;
```
Comment (david-mears-2) [inline src/controllers/indexController.ts:5]: How does this value get into the process.env object?

File: src/logging.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    const customFormat = (
+        tokens: Dict<(req: Request, res: Response, header?: string) => string>,
```
Comment (david-mears-2) [inline src/logging.ts:7]: Not able to parse this - what does 'tokens' refer to?

**Review by david-mears-2 — APPROVED** [review]
It works for me. I have questions, and a comment about ConfigReader.

File: src/configReader.ts (hunk 3 above)
Comment (EmmaLRussell) [reply src/configReader.ts:11]: Debatable I think. This class was largely ripped off from WODIN where we certainly does have multiple config files. We *may* end up with multiple config files in grout, I think it's still tbd. 

Either way I think it's ok for the server module to centralise knowledge of where all the folders are. That's the approach I've been taking anyway!

File: src/controllers/indexController.ts (hunk 4 above)
Comment (EmmaLRussell) [reply src/controllers/indexController.ts:5]: Through the magic of running from an `npm` script :)
https://docs.npmjs.com/cli/v9/using-npm/scripts#packagejson-vars

File: src/logging.ts (hunk 5 above)
Comment (EmmaLRussell) [reply src/logging.ts:7]: It refers to the bits of information relating to a request/response pair which can be logged by `morgan` (the logging package). As you can see from the type, these are actually functions, and this whole `customFormat` method is passed to `morgan` to tell it how to log requests.

Comment (david-mears-2) [reply src/logging.ts:7]: It's confusing, but I guess all is explained in the morgan docs.

**Review by absternator — APPROVED** [review]

**Pull request merged**